From 0d8b8ba2b876a7990bbbe5a618925598bd07363a Mon Sep 17 00:00:00 2001
From: Luciano Coelho <coelho@ti.com>
Date: Fri, 30 Sep 2011 16:20:04 +0300
Subject: [PATCH 09/29] wpa_s: add filter support to scheduled scans

Pass SSIDs to be matched in scheduled scan results.  Only the SSIDs
that are included in the match lists will be reported by the driver,
so the filtering can be offloaded to the hardware and the power
consumption can be reduced.

Signed-off-by: Luciano Coelho <coelho@ti.com>
---
 src/drivers/driver.h              |    1 +
 wpa_supplicant/scan.c             |   35 +++++++++++++++++++++--------------
 wpa_supplicant/wpa_supplicant.c   |    1 +
 wpa_supplicant/wpa_supplicant_i.h |    1 +
 4 files changed, 24 insertions(+), 14 deletions(-)

diff --git a/src/drivers/driver.h b/src/drivers/driver.h
index bffa0bb..9fe25f2 100644
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -583,6 +583,7 @@ struct wpa_driver_capa {
 	int max_scan_ssids;
 	int max_sched_scan_ssids;
 	int sched_scan_supported;
+	int max_match_sets;
 
 	/**
 	 * max_remain_on_chan - Maximum remain-on-channel duration in msec
diff --git a/wpa_supplicant/scan.c b/wpa_supplicant/scan.c
index a4a5bb3..40a97e1 100644
--- a/wpa_supplicant/scan.c
+++ b/wpa_supplicant/scan.c
@@ -597,7 +597,6 @@ int wpa_supplicant_req_sched_scan(struct wpa_supplicant *wpa_s)
 	enum wpa_states prev_state;
 	struct wpa_ssid *ssid;
 	int ret;
-	int use_wildcard = 0;
 	int max_sched_scan_ssids;
 
 	if (!wpa_s->sched_scan_supported)
@@ -615,6 +614,10 @@ int wpa_supplicant_req_sched_scan(struct wpa_supplicant *wpa_s)
 
 	os_memset(&params, 0, sizeof(params));
 
+	/* if we can't allocate space for the filters, we just don't filter */
+	params.filter_ssids = os_zalloc(wpa_s->max_match_sets *
+					sizeof(struct wpa_driver_scan_filter));
+
 	prev_state = wpa_s->wpa_state;
 	if (wpa_s->wpa_state == WPA_DISCONNECTED ||
 	    wpa_s->wpa_state == WPA_INACTIVE)
@@ -649,39 +652,43 @@ int wpa_supplicant_req_sched_scan(struct wpa_supplicant *wpa_s)
 			continue;
 		}
 
-		if (!ssid->scan_ssid)
-			use_wildcard = 1;
-		else {
+		if (params.filter_ssids && ssid->ssid && ssid->ssid_len) {
+			os_memcpy(params.filter_ssids[params.num_filter_ssids].ssid,
+				  ssid->ssid, ssid->ssid_len);
+			params.filter_ssids[params.num_filter_ssids].ssid_len =
+				ssid->ssid_len;
+			params.num_filter_ssids++;
+		}
+
+		if (ssid->scan_ssid) {
 			params.ssids[params.num_ssids].ssid =
 				ssid->ssid;
 			params.ssids[params.num_ssids].ssid_len =
 				ssid->ssid_len;
 			params.num_ssids++;
-			if (params.num_ssids + 1 >= max_sched_scan_ssids) {
+			if (params.num_ssids >= max_sched_scan_ssids) {
 				wpa_s->prev_sched_ssid = ssid;
 				break;
 			}
 		}
+
+		if (params.num_filter_ssids >= wpa_s->max_match_sets)
+			break;
 		wpa_s->prev_sched_ssid = ssid;
 		ssid = ssid->next;
 	}
 
-	if (ssid || use_wildcard) {
-		wpa_dbg(wpa_s, MSG_DEBUG, "Include wildcard SSID in "
-			"the sched scan request");
-		params.num_ssids++;
-	} else {
-		wpa_dbg(wpa_s, MSG_DEBUG, "ssid %p - list ended", ssid);
-	}
-
-	if (!params.num_ssids)
+	if (!params.num_ssids) {
+		os_free(params.filter_ssids);
 		return 0;
+	}
 
 	wpa_dbg(wpa_s, MSG_DEBUG, "Starting sched scan: interval %d timeout %d",
 		wpa_s->sched_scan_interval, wpa_s->sched_scan_timeout);
 
 	ret = wpa_supplicant_start_sched_scan(wpa_s, &params,
 					      wpa_s->sched_scan_interval);
+	os_free(params.filter_ssids);
 	if (ret) {
 		wpa_msg(wpa_s, MSG_WARNING, "Failed to initiate sched scan");
 		if (prev_state != wpa_s->wpa_state)
diff --git a/wpa_supplicant/wpa_supplicant.c b/wpa_supplicant/wpa_supplicant.c
index 4269ef0..c354b5f 100644
--- a/wpa_supplicant/wpa_supplicant.c
+++ b/wpa_supplicant/wpa_supplicant.c
@@ -2267,6 +2267,7 @@ next_driver:
 		wpa_s->max_scan_ssids = capa.max_scan_ssids;
 		wpa_s->max_sched_scan_ssids = capa.max_sched_scan_ssids;
 		wpa_s->sched_scan_supported = capa.sched_scan_supported;
+		wpa_s->max_match_sets = capa.max_match_sets;
 		wpa_s->max_remain_on_chan = capa.max_remain_on_chan;
 		wpa_s->max_stations = capa.max_stations;
 	}
diff --git a/wpa_supplicant/wpa_supplicant_i.h b/wpa_supplicant/wpa_supplicant_i.h
index 4f9c19a..c96ee9b 100644
--- a/wpa_supplicant/wpa_supplicant_i.h
+++ b/wpa_supplicant/wpa_supplicant_i.h
@@ -428,6 +428,7 @@ struct wpa_supplicant {
 	int max_scan_ssids;
 	int max_sched_scan_ssids;
 	int sched_scan_supported;
+	int max_match_sets;
 	unsigned int max_remain_on_chan;
 	unsigned int max_stations;
 
-- 
1.7.0.4

